### ISSUE TITLE
Safety margins value getter

### ISSUE THREAD

**Issue by dirkcremers** [capacitor-community/safe-area#8]
First of all, thanks for maintaining the plugin. It's really helpful, and I appreciate your effort.

**Is your feature request related to a problem? Please describe.**
After integrating the plugin and configuring settings in capacitor.config.ts, our application achieved an edge-to-edge display, which is fantastic. However, we're encountering issues with content overlapping with the device notch, for instance. We're exploring options to maintain edge-to-edge layout while ensuring content remains within safety margins.

**Describe the solution you'd like**
We're looking for a getter function that can provide the safety margin values for different devices.

**Describe alternatives you've considered**
-

**Additional context**
As I'm new to Capacitor, I'm still familiarizing myself with its framework. However, I'm eager to learn and contribute to the project in case I can.

**Comment by frederikheld** [comment]
There is a different plugin that does this: https://www.npmjs.com/package/capacitor-plugin-safe-area

I've never used it though, so no recommendation, just information.

**Comment by tafelnl** [comment]
Could you provide me with some info as why you would want a getter function? This plugin injects a CSS variable which you can use directly in your CSS. For example, it enables you to do the following:

```css
#header {
  padding-top: var(--safe-area-inset-top);
}
```

Maybe it's not very clear from the docs that this is the case. So I will update the docs accordingly. Please let me know if this 'fixes' your issue

**Comment by tafelnl** [comment]
@frederikheld Thank you for the information. If you'd like to know the differences between that plugin and this one, I've outlined them here: https://github.com/AlwaysLoveme/capacitor-plugin-safe-area/issues/29

**Comment by tafelnl** [comment]
I think the docs are a bit easier to understand with this change: #11

Let me know what you think and if your issue is resolved now

**Comment by frederikheld** [comment]
Thank you for the hint about the injected css variable. A JavaScript getter would still be useful.

Use Case:

I have a map that covers the top half of the screen and goes to right through the edge. So there is an unsafe part that the user can't interact with. The map is centered to a location and I want the centering to compensate for the safe area.

**Comment by frederikheld** [comment]
> Could you provide me with some info as why you would want a getter function? This plugin injects a CSS variable which you can use directly in your CSS. For example, it enables you to do the following:
> 
> ```css
> #header {
>   padding-top: var(--safe-area-inset-top);
> }
> ```
> 
> Maybe it's not very clear from the docs that this is the case. So I will update the docs accordingly. Please let me know if this 'fixes' your issue

A question about this: did I understand the whole safe area CSS stuff correctly that `env(safe-area-inset-*)` only exists on iOS, but your solution `var(--safe-area-inset-top)` works on all platforms?

**Comment by tafelnl** [comment]
`env(safe-area-inset-*)` exists on all platforms (https://developer.mozilla.org/en-US/docs/Web/CSS/env#browser_compatibility). But there's a bug on Android that it doesn't provide the correct values when trying to use it in a webview with Edge-to-Edge enabled

**Comment by frederikheld** [comment]
Ah, okay. Thanks a lot for the clarification!

**Comment by tafelnl** [comment]
As for your other question, I'm not sure I'm willing to accommodate for that type of behavior in this plugin, for several reasons:

- I want to keep this plugin as simple as possible
- It's another feature to maintain
- Most importantly, it's already possible in plain JavaScript: https://stackoverflow.com/a/68526312/8634342

**Comment by frederikheld** [comment]
I understand. But reading computed styles only works in some cases. If the plugin knows those values before the CSS is rendered, it would help if we could access them directly through JS.

**Comment by tafelnl** [comment]
> If the plugin knows those values before the CSS is rendered, it would help if we could access them directly through JS.

Using a `new MutationObserver` instead of a `SafeArea.addListener('safeAreaChanged)` for example, is just as fast. They both have to wait till the JavaScript has been loaded.

> But reading computed styles only works in some cases.

I'm not sure what you mean by this

**Comment by tafelnl** [comment]
> I have a map that covers the top half of the screen and goes right to the edge. So there is an unsafe part that the user can't interact with and that will be overlaid with status bar text and icons. The map is centered to a location and I want the centering to compensate for the unsafe area.

How is this not already possible using CSS by the way?

**Comment by frederikheld** [comment]
> > If the plugin knows those values before the CSS is rendered, it would help if we could access them directly through JS.
> 
> Using a `new MutationObserver` instead of a `SafeArea.addListener('safeAreaChanged')` for example, is just as fast. They both have to wait till the JavaScript has been loaded.
> 
> > But reading computed styles only works in some cases.
> 
> I'm not sure what you mean by this

`computed style` means: the CSS values that the browser has computed after applying all CSS directives. The browser computes those values while rendering. So they will only be available AFTER rendering. But in some cases they are needed before rendering because otherwise you have to update an element that has already been rendered which can cause layout shifts.

This is a common issue when you use `getComputedStyle` to read the hight of an element with dynamic height because the height will only be known after the element has been rendered and `display` and `position` attributes have been applied.

I expect the same issue here.

**Comment by frederikheld** [comment]
> > I have a map that covers the top half of the screen and goes right to the edge. So there is an unsafe part that the user can't interact with and that will be overlaid with status bar text and icons. The map is centered to a location and I want the centering to compensate for the unsafe area.
> 
> How is this not already possible using CSS by the way?

Because the Google Maps JS SDK is a very well designed modern library with great UX /s IDK, ask the Google guys. But you do the centering via JS. Of course I could move the whole map via CSS, but then I don't have the e2e look anymore. I just want to move the center to have the map centered within the safe area.

**Comment by tafelnl** [comment]
Understandable. To support this use case, I would have to write logic for web and iOS as well. Which I'm deliberately trying to prevent - as opposed to the other solutions available. So I'm still not very eager to implement such functionality.

Also it's not possible to listen to safe area inset changes on web apparently: https://github.com/w3c/csswg-drafts/issues/2732, https://github.com/w3c/csswg-drafts/issues/2628

So you will have listen to screen size changes or something. This is not something I'm willing to get into. It will probably get messy and that's not something I'm looking forward to maintain. Especially since probably most users won't need it

**Comment by frederikheld** [comment]
Ok, understandable. Would you accept a PR? I won't promise anything, but if I don't find a solution to my case, I might be looking into it.

**Comment by tafelnl** [comment]
Yeah sure, go ahead! But to be honest, I can't give you any guarantees I will be merging it. It also depends a bit on the solution you end up finding and implementing. If it's - purely hypothetically - like 10 lines of code, it's a lot more likely for me to accept it and maintain it, than when it's like a few hundred lines of code.

**Comment by dirkcremers** [comment]
Hi, sorry for the slow response. Just came back from holiday!

> There is a different plugin that does this: https://www.npmjs.com/package/capacitor-plugin-safe-area
> 
> I've never used it though, so no recommendation, just information.

Yeah true. However, people were complaining about the compatibility of the plugin with regards to capacitor v6.

So basically, my app now has an edge-to-edge design, which looks great. However, I'd like to add some margin or padding to the interactive content (e.g., buttons) to prevent them from overlapping with the notch. Currently, in the edge-to-edge layout, this issue occurs. It's essential that the interactive content remains clickable. This might align with Frederikheld's comments.

**Comment by frederikheld** [comment]
@dirkcremers This is exactly what @capacitor-community/safe-area helps you with. It exposes the correct margins as CSS values so you can use them to set margins to your containers. @tafelnl updated the [README.md](https://github.com/capacitor-community/safe-area/blob/master/README.md) to make this more clear.

My use case is a different one where I need the value in JavaScript. But yours sounds like it can be solved with pure CSS.

**Comment by dirkcremers** [comment]
Thanks for both your replies! I will try it out

**Comment by efleetcon-simon-s** [comment]
For iOS, having content below the notch or dynamic island would be bad UI. The safe are insets help to offset a view's content below. When we write code for multiple platforms (e.g. web & iOS), we will end up having no space above the view at all on web, which again looks cluttered. If I now add additional margin on web, that will also be added to the one already applied by using the safe area insets on iOS, causing a larger, unnatural gap on the platform. I cannot apply conditional styles based on a CSS variable (except with some hacky `min()` `max()` trickery, which leads me to either having a larger than intended gap on iOS, or no gap at all on the web. If the values were accessible in JS (like they are in [this](https://github.com/AppAndFlow/react-native-safe-area-context) package, I could render a gap only if no safe area inset given. Rendering the gap based on the platform will not work because we have iOS devices without notch. Or without bottom bar (older ones with home buttons).

**Comment by frederikheld** [comment]
We do understand how the plugin works . It's just not enough to handle all issues that come with safe are insets. The plugin only covers the CSS side of the solution. It would be good to have the JS side covered as well because a web app usually has both.
